actors/builtin/miner/miner_state.go

```diff
@@ -230,11 +230,36 @@ func (st *State) IsSectorInTemporaryFault(store adt.Store, sectorNo abi.SectorNu
 	return ret, nil
 }
 
-func (st *State) ComputeProvingSet() cid.Cid {
+func (st *State) ComputeProvingSet(store adt.Store) (cid.Cid, error) {
 	// Current ProvingSet is a snapshot of the Sectors AMT subtracting sectors in the FaultSet
-	// TODO: actually implement this
-	var ret cid.Cid
-	return ret
+
+	provingSet, err := adt.MakeEmptyArray(store)
+	if err != nil {
+		return provingSet.Root(), errors.Wrapf(err, "failed to create a new provingSet")
+	}
+
+	var err2 error
+	err = st.forEachSector(store, func(sector *SectorOnChainInfo) {
+		fault, err := st.FaultSet.Has(uint64(sector.Info.SectorNumber))
+		AssertNoError(err)
+		Assert(sector.DeclaredFaultEpoch != epochUndefined)
+		Assert(sector.DeclaredFaultDuration != epochUndefined)
+
+		if !fault {
+			err2 := provingSet.Set(uint64(sector.Info.SectorNumber), sector)
+			if err2 != nil {
+				return
+			}
+		}
+	})
```

Comment:
Thanks for this. I think it's complicated™️ . 

When we talked, I think I'd indicated that there was no need to subtract the fault set from this, just copy the sectors root CID. Thanks for realising that we can't lazily subtract the faultset during evaluation, because the fault set might have changed. But:
- If we're just trying to record the IDs of the sectors in the proving set, we should use the much more compact bitfield representation rather than construct a new AMT. The lightweight clone approach of copying the root CID is only lightweight if the collection is identical. This will write a whole new AMT each proving period, which is quite expensive.
- But but, the values in this AMT are SectorInfos, which are mutable objects. So by cloning the AMT in this way we're actually baking in the present fault information in the form of the DeclaredFaultEpoch/Duration fields. So actually the faults are frozen after all so it *is* safe to just do the cheap copy (but have to traverse it to figure out the sector id set).. But only as fallout from the otherwise unfortunate design of having mutable sector state on chain.
- But but but, `st.ProvingSet` appears to be unused in the miner actor! So I can't tell what the more appropriate representation is to make a recommendation here. I can see a thread of spec debt where `PoStVerifyInfo.EligibleSectors` is unimplemented, a Go map that should almost certainly be something else (it can't be a H/AMT). Probably that's supposed to be a bitfield, which would then suggest option 1 above.

@magik6k can you confirm that approach?

Reply:
Thanks for the catch. Just updated to call `st.ProvingSet` appropriately, @anorth. Sorry about that.

Thanks also for spelling out the above.

Your note does make me wonder why we even include a `FaultSet` in the first place? Is the hope to do easy-lookup of faults? Given that `IsSectorInTemporaryFault()` gets the sector as a whole, why not simply read its `DeclaredFaultEpoch` and `DeclaredFaultDuration` there and save ourselves a bitfield?

To repeat what you said above though (and make sure I get it), you are saying that given sectors include the fault fields, doing the AMT clone (simply copying the root CID of `sectors` as the`provingSet`) and then later traverse it to get only those IDs we want (you could imagine a `SetProvingSet()` method which does the snapshot while `ComputeProvingSet()` pares it down) should work given it will be a snapshot of those sectors as they were at that time (given further mutation will lead to a new root). Is that right?